runtime/gc_base/ObjectAccessBarrier.cpp

```diff
@@ -1546,7 +1574,22 @@ MM_ObjectAccessBarrier::copyObjectFields(J9VMThread *vmThread, J9Class *objectCl
 		while (offset < limit) {
 			UDATA srcAddress = (UDATA)srcObject + srcOffset + offset;
 			UDATA destAddress = (UDATA)destObject + destOffset + offset;
-			if (sizeof(uint32_t) == referenceSize) {
+			bool copy64Bits = false;
+			if (isValueType
+				&& (sizeof(uint32_t) == referenceSize)
+				&& (0 == (srcAddress & 7))
+				&& ((offset + referenceSize) < limit)
+			) {
+				copy64Bits = true;
+			}
+			if (copy64Bits) {
+				*(uint64_t *)destAddress = *(uint64_t *)srcAddress;
```

Comment:
I don't understand how this makes sense because `copy64Bits` implies `sizeof(uint32_t) == referenceSize`.

Reply:
It comes from https://github.com/eclipse-openj9/openj9/issues/11178#issuecomment-740726529.

On 32-bit reference mode, we do not want to tear 64-bit field into 2 copies. For 64-bit field, they are properly aligned, so we can do 64-bit copy. For 32-bits fields, we can also do 64-bit copy (to copy one 32-bit filed or two 32-bit fileds), as the whole object is going to be copied anyway.

Reply:
I suggest that loop could be improved so there are only two cases to consider each iteration instead of three:
```
	while (offset < limit) {
		UDATA srcAddress = (UDATA)srcObject + srcOffset + offset;
		UDATA destAddress = (UDATA)destObject + destOffset + offset;
		/* prefer to copy 64 bits at a time if possible */
		if ((sizeof(uint64_t) == referenceSize)
			|| (isValueType && (0 == (srcAddress & 7)) && ((offset + referenceSize) < limit))
		) {
			*(uint64_t *)destAddress = *(uint64_t *)srcAddress;
			offset += sizeof(uint64_t);
		} else {
			*(uint32_t *)destAddress = *(uint32_t *)srcAddress;
			offset += sizeof(uint32_t);
		}
	}
```
A further improvement would recognize that it copies 0 or 1 32-bit words, followed by n 64-bit words and finally 0 or 1 32-bit words.

Reply:
> I suggest that loop could be improved so there are only two cases to consider each iteration instead of three

I have modified this loop as suggested.